### PR TITLE
add support for TS 5.0

### PR DESCRIPTION
**Pull request by AlCalzone**
Almost done. The `canPatch` check needs fixing and `patchModuleDiagnostics` is broken for TS 5.0.

### CODE REVIEW

File: src/installer/lib/file-utils.ts [hunk 1]
```diff
@@ -136,7 +136,7 @@ export function getTSModule(file: string, includeSrc: boolean = false): TSModule
   const filename = path.basename(file);
   const dir = path.dirname(file);
   const fileData = fs.readFileSync(file, 'utf8');
-  const canPatch = Boolean(fileData.match(/^\(function\s\(ts\)\s?{[\s\S]+?\(ts\s?\|\|\s?\(ts\s?=\s?{}\)\);?$/m));
+  const canPatch = true; // Boolean(fileData.match(/^\(function\s\(ts\)\s?{[\s\S]+?\(ts\s?\|\|\s?\(ts\s?=\s?{}\)\);?$/m));
```
Comment (AlCalzone) [inline src/installer/lib/file-utils.ts:139]: This was a test to see how far I get. Could use a hint how to test this on v5.

File: src/installer/lib/patcher.ts [hunk 2]
```diff
@@ -95,6 +96,10 @@ export function patchTSModule(tsModule: TSModule, tsPackage: TSPackage) {
   const isTSC = (filename === 'tsc.js');
   const patchSrc = generatePatch(isTSC);
 
+  const ver = tsPackage.version.split('.')
+  const majorVer = +ver[0];
+  const minorVer = +ver[1];
+
   /* Add diagnostic modification support */
   const moduleSrc = patchModuleDiagnostics(tsModule, tsPackage);
```
Comment (AlCalzone) [inline src/installer/lib/patcher.ts:104]: This throws when attempting to patch in v5.
I also don't understand why we seem to patch the module diagnostics twice, once here ...

File: src/installer/lib/patcher.ts [hunk 3]
```diff
@@ -155,3 +143,76 @@ export function patchTSModule(tsModule: TSModule, tsPackage: TSPackage) {
...
+  const execCmdPost = tscSlice.lastIndexOf('\nts.executeCommandLine\(');
+  if (execCmdPost < 0)
+    throw new Error(`Could not find tsc executeCommandLine`);
+  const execCmd = tscSlice.slice(execCmdPost);
+  tscSlice = tscSlice.slice(0, execCmdPost);
+
+  /* Expand TSC with full typescript library (splice tsc part on top of typescript.ts code) */
+  return [
+    getHeader(),
+    patchModuleDiagnostics(tsModule, tsPackage, fs.readFileSync(tsFile, 'utf-8')),
```
Comment (AlCalzone) [inline src/installer/lib/patcher.ts:160]: ... and again here.

**Comment by AlCalzone** [comment]
Problems with the current patch:
1. The `__export` helper function (this one is duplicated btw) must be changed, so that it sets the `configurable` flag. Otherwise we are not able to overwrite `createProgram`
```diff
-      __defProp(target, name, { get: all[name], enumerable: true });
+      __defProp(target, name, { get: all[name], enumerable: true, configurable: true });
```
2. Inside `tsc.js`, there is no `ts` variable anymore. At the moment, it's called `ts_exports3`, but I suppose that can and will change. Fortunately, we should be able to find it with a text search:
```js
  // src/typescript/_namespaces/ts.ts
  var ts_exports3 = {};

  __export(ts_exports3, {
```
and then either use that instead of `ts` in the patch, or alias it in the `tsp` closures.

File: src/installer/lib/patcher.ts [hunk 4]
```diff
@@ -13,7 +14,7 @@ const dtsPatchSrc = '\n' +
   fs.readFileSync(path.resolve(appRoot, tspPackageJSON.directories.resources, 'module-patch.d.ts'), 'utf-8');
 
 const jsPatchSrc =
-  fs.readFileSync(path.resolve(appRoot, tspPackageJSON.directories.resources, 'module-patch.js'), 'utf-8')
+  fs.readFileSync(path.resolve(appRoot, tspPackageJSON.directories.resources, 'module-patch-v5.js'), 'utf-8')
```
Comment (AlCalzone) [inline src/installer/lib/patcher.ts:17]: this needs to be dynamic

File: src/patch_v5/lib/plugin.ts [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+          ret = (factory as TypeCheckerPattern)(program.getTypeChecker(), cleanConfig);
+          break;
+
+        case undefined:
+        case 'program':
+          const { addDiagnostic, removeDiagnostic, diagnostics } = diagnosticExtrasFactory(program);
+
+          ret = (factory as ProgramPattern)(program, cleanConfig, {
+            // @ts-expect-error TS internally renames its exports. TODO: Find the correct name, or alias it somehow
+            ts: ts_exports3,
```
Comment (AlCalzone) [inline src/patch_v5/lib/plugin.ts:224]: I don't like this

**Comment by bradennapier** [comment]
Any update here?

**Comment by nonara** [comment]
@bradennapier Problem is solved and it's looking good. We're cleaning up approach and finishing some things. Expecting to be released next week.

**Comment by bradennapier** [comment]
Should have released earlier as beta ;-) we are eager to get to ts 5.0 :-P but thanks!

**Comment by nonara** [comment]
It isn't complete enough to do that yet, but if there's a delay for adding tests, etc., I'm planning on releasing a beta.

**Pull request merged**